### PR TITLE
Flush stdout when control returns to shell

### PR DESCRIPTION
**Pull request by mqudsi**
This prevents kernel-buffered stdout contents from being written to the terminal after the prompt itself, which can happen when the external process is terminated with a signal and hasn't executed the libc/runtime-provided exit handlers that flush stdout.

Screenshots of afl-fuzz being terminated by `^C` before and after the change:

Before:

image

After:

image

I tried to minimize any performance impact by trying to reduce the number of times this is called (by placing it as high up as possible, but bubbling up reap status) and by not performing the call in non-interactive mode (where fish shouldn't be painting to the shell bypassing stdout). stderr isn't flushed because it's typically unbuffered.

Let me know if you think this should be included in the changelog afterwards.

## TODOs:

- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.rst

### CODE REVIEW

File: src/proc.rs [hunk 1]
```diff
@@ -1066,6 +1067,12 @@ impl Job {
                     parser.libdata_mut().status_count += 1;
                 }
             }
+
+            // Prevent kernel-buffered output from being written to the terminal after the prompt.
+            // We don't need to this if we're running non-interactively, where no prompt is shown.
+            if external_proc_exited && parser.is_interactive() {
+                let _ = std::io::stdout().flush();
```
Comment (krobelus) [inline src/proc.rs:1074]: Seems reasonable.
Maybe we should do this only if the job is `signal_exited`?
I don't know if this scenario is also possible on normal exit.

Another approach is to do it in `Reader::readline()` before `exec_prompt`,
using something like `exec_external_count` to decide whether we want to flush.

(Both approaches won't handle the case when the output comes from a background job that's still running).

> Let me know if you think this should be included in the changelog afterwards.

Good question; can go both ways I think.
I don't think there is an obligation to add it

**Comment by zanchey** [comment]
I'd mention it:

"Standard output is flushed after external commands exit, in an attempt to reduce interleaving their output with the prompt (:issue:`10861`)."

File: src/proc.rs (hunk 1 above)
Comment (mqudsi) [reply src/proc.rs:1074]: I think keeping the flush call closer to the exec code makes more sense than putting it in the reader, btw.

**Review by ridiculousfish — CHANGES_REQUESTED** [review]
This doesn't make any sense to me. In my understanding, `std::io::stdout().flush()` flushes Rust's *internal* buffer. Because this is in-process, there is no way for an external command, such as afl-fuzz, to append to it, so there is nothing to flush, so I would not expect this to do anything at all.

If there is something to flush, then it ought to have been flushed before the external command ran!

Can you please write a command line program that demonstrates this in action? afl-fuzz doesn't seem available on Apple Silicon.

**Comment by mqudsi** [comment]
You're right; I think I ran into a change in timing that resulted in the "corrected" behavior. I was thinking in terms of `fsync()` rather than `flush()` but that doesn't really mean anything for a fifo buffer. I really don't know why I didn't catch this, thank you for bringing me to my senses.

Are race conditions here inevitable?

**Comment by ridiculousfish** [comment]
I think *some* race conditions are inevitable, especially with curses-style apps such as afl-fuzz. For example, if the app moves the cursor to some point on the screen and then dies, then fish will inherit a weird cursor position and I can see how that might screw up the prompt - though I would expect the `PROMPT_SP` hack to handle that (odd that it doesn't).

If you want to pursue this, it would be useful to write a custom tool that reproduces this, so we can explore more.

**Comment by mqudsi** [comment]
It actually reproduces when afl is still starting up and only producing a colorized log of text to stdout with no real tui to speak of, but it must be using absolute positioning ansi escapes or similar because regular stdout is strictly linear. 

I wonder if I can just record and replay the output of afl instead of trying to independently reproduce the race.

**Comment by krobelus** [comment]
`script` from util-linux has a `--log-timing` option that should enable almost-perfect replay

**Comment by mqudsi** [comment]
Thanks, that's useful. Probably have to use socat to trick it into thinking its running in a terminal, though.

**Comment by mqudsi** [comment]
I wrote this last week but never hit submit:

-------------


I didn't need to use `socat` as `script` uses a pseudotty itself.

Here's a sample (replay with `scriptreplay --timing=timing.log cargo-afl.log`) that reproduces the issue: [cargo-afl.log](https://github.com/user-attachments/files/17894249/cargo-afl.log), [timing.log](https://github.com/user-attachments/files/17894250/timing.log)

Looking at the output with `sequin` makes it easier to understand all the output, but it's not particularly elucidating. There doesn't seem to be any absolute positioning of text, in fact it just looks like fish took over but `cargo afl` kept writing, probably via a child process that inherited the same tty.

One obvious answer is to set `+tostop` at all times except when fish has explicitly launched a background process, but there are obvious drawbacks.

**Pull request closed without merging**